pipeline/lib/custom_dockerfile.ml

```diff
@@ -50,92 +50,40 @@ let get_all_files ~repository =
-        libpcre3-dev && opam remote add origin https://opam.ocaml.org && opam \
-        update"
+       "sudo apt-get install -qq -yy --no-install-recommends libffi-dev \
+        liblmdb-dev m4 pkg-config libgmp-dev libssl-dev libpcre3-dev"
   @@ run "sudo mv /usr/bin/opam-2.1 /usr/bin/opam"
-
-let add_workdir =
-  let open Dockerfile in
-  copy ~src:[ "--chown=opam:opam ." ] ~dst:"bench-dir" ()
+  @@ run "opam remote add origin https://opam.ocaml.org"
+  @@ run "opam update"
   @@ workdir "bench-dir"
-  @@ add ~src:[ "--chown=opam ." ] ~dst:"." ()
+  @@ copy ~src:[ "--chown=opam:opam ./*.opam" ] ~dst:"." ()
   @@ run "opam exec -- opam pin -y -n --with-version=dev ."
+  @@ run "opam exec -- opam install -y --depext-only ."
+  @@ run "opam exec -- opam install -y --deps-only ."
+  @@ copy ~src:[ "--chown=opam:opam ." ] ~dst:"bench-dir" ()
+  @@ add ~src:[ "--chown=opam ." ] ~dst:"." ()
 
-let minimal_dockerfile ~base =
-  let open Dockerfile in
-  base_dockerfile ~base @@ add_workdir
-
-let docker_exec ~label ~pool ~run_args ~repository ~dockerfile args =
-  let { Repository.branch; pull_number; _ } = repository in
-  let repo_info = Repository.info repository in
-  let src = Repository.src repository in
-  let image = Docker.build ~pool ~pull:false ~dockerfile (`Git src) in
-  let commit = Repository.commit_hash repository in
-  Current_util.Docker_util.pread_log ~label ~pool ~run_args ~repo_info
-    ?pull_number ?branch ~commit ~args image
-
-let opam_install ~opam_file =
-  match String.compare opam_file "" with
-  | 0 -> Format.sprintf "opam exec -- opam install -y --deps-only ."
-  | _ ->
-      Format.sprintf "opam exec -- opam install %s -y --deps-only ." opam_file
-
-let weekly = Current_cache.Schedule.v ~valid_for:(Duration.of_day 7) ()
-
-let base = Docker.pull ~schedule:weekly "ocaml/opam"
-
-let with_base f = Current.map (fun base -> `Contents (f ~base)) base
-
-let discover_dependencies ~pool ~run_args ~repository ~opam_file =
-  let cmd =
-    opam_install ~opam_file
-    ^ " --dry-run | sed -n '/^Installing \\(.*\\).$/{s//\\1/g;p}'"
-  in
-  let args = [ "/bin/sh"; "-c"; cmd ] in
-  let dockerfile = with_base minimal_dockerfile in
-  let+ output =
-    docker_exec ~label:"discover-dependencies" ~pool ~run_args ~repository
-      ~dockerfile args
-  in
-  String.concat " " (String.split_on_char '\n' output)
-
-let dockerfile ~base ~dependencies ~opam_file =
-  let open Dockerfile in
-  let install_static_dependencies =
-    if dependencies = "" then empty
-    else run "opam install -y %s || exit 0" dependencies
-  in
-  base_dockerfile ~base
-  @@ install_static_dependencies
-  @@ add_workdir
-  @@ run "%s" (opam_install ~opam_file)
-
-let dockerfile ~pool ~run_args ~repository ~opam_file =
-  let* dependencies =
-    discover_dependencies ~pool ~run_args ~repository ~opam_file
-  in
-  with_base (dockerfile ~dependencies ~opam_file)
+let dockerfile ~base = Dockerfile.crunch (dockerfile ~base)
```

Comment:
What does `crunch` do? :)

Reply:
`Dockerfile.crunch` collapses multiple docker `run` commands into a single one, using `&&` in between (so that the docker cache a bit smaller). It's a prettier alternative to writing long commands in multiline strings :P

Reply:
nice!! very cool!